docs/proposals/maintenance.md

```diff
@@ -0,0 +1,179 @@
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+<strong>
+The latest release of this document can be found
+[here](http://releases.k8s.io/release-1.1/docs/proposals/maintenance.md).
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Node Maintenance
+
+## Abstract
+
+A proposal for implementing safe node maintenance primitives, in stages.
+
+## Overview
+
+Nodes need periodic maintenance, most of which require partial or total
+unavailability to pods.  The cluster admin would like for this to happen (1)
+automatically, and (2) respecting any service quality constraints.
+
+Examples of maintenance:
+
+- unmount/remount filesystems to change mkfs params
+- remove from service but leave up for debugging
+- reboot for kernel or image upgrade
+- remove from service for hardware repair
+- remove from service permanently
+
+## Service Quality
+
+This may be thought of as the cell admin offering a guarantee or SLO, or as the
+user specifying their requirements.  Around evictions, there are typically three
+parameters:
+
+- time between evictions (of pods in the same replica set)
+- minimum shard strength (minimum fraction of pods healthy)
+- eviction notice
+
+## Implementations
+
+In order of increasing complexity (and benefit).
+
+### Drain one node
+
+This is implemented in https://github.com/kubernetes/kubernetes/pull/16698
+
+### One-shot drain of multiple nodes
+
+In this case, we try to meet some specific service goals specified by the admin
+on the command line.  The specification of these parameters only when we run the
+tool lets us provide the possibility of safety while avoiding questions of where
+to configure and store such specification (treated further down).
+
+```
+kubectl drain \
+  --min-shard-strength=0.75 \
+  --min-seconds-between-evictions=900 \
+  --grace=900 $nodes
+```
+
+This would still be pretty simple.  We offer no guarantees about safety in the
+face of crashes or restarts, but we do meet the needs of admins who want
+something more sophisticated than `sleep`.
+
+### Implementation of the above as a controller
+
+The parameters of a specific lame, drain, or unlame request could be pushed to
+the API server as a drain object.  The drain controller would have to manage
+multiple drains in flight at once.  If the drains have differing parameters,
+then more aggressive drains would generally starve safer ones.  While this could
+be a useful feature in emergencies, it will be a lot easier to reason about
+multiple drains in progress if they all share the same parameters, which would
+be configured via kube-env and would represent the cell admin's SLO.
+
+The drain object would thus look something like this
+
+```
+type DrainSpec struct {
+  LameNodes    PodList
```

Comment:
On Wed, Nov 18, 2015 at 10:24 PM, Paul Morie <notifications@github.com>
wrote:

> In docs/proposals/maintenance.md
> <https://github.com/kubernetes/kubernetes/pull/17393#discussion_r45306233>
> :
>
> > +
> > +### Implementation of the above as a controller
> > +
> > +The parameters of a specific lame, drain, or unlame request could be pushed to
> > +the API server as a drain object.  The drain controller would have to manage
> > +multiple drains in flight at once.  If the drains have differing parameters,
> > +then more aggressive drains would generally starve safer ones.  While this could
> > +be a useful feature in emergencies, it will be a lot easier to reason about
> > +multiple drains in progress if they all share the same parameters, which would
> > +be configured via kube-env and would represent the cell admin's SLO.
> > +
> > +The drain object would thus look something like this
> > +
> > +```
> > +type DrainSpec struct {
> > +  LameNodes    PodList
>
> Suggest trying to find a different name than lame, which might have
> connotations that are undesirable. How about referring to this state as
> reduced-availability?
>

I would like a term that cleanly has two verbs and a noun describing the
state.  Do you have any suggestions along those lines?  In the current
case: lame, unlame, and lamed.

Some suggestions from discussions with @bgrant0607 and @erictune

- cordon, uncordon, cordoned
- decommission, recommission, decommissioned
- close, open, closed (as in a faucet or business)
- shut, reopen, shut (ditto)

—
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/17393/files#r45306233>.
>